internal/item/item.go

```diff
@@ -0,0 +1,269 @@
+package item
+
+import (
+	"errors"
+	"time"
+
+	"github.com/ardanlabs/kit/db"
+	"github.com/ardanlabs/kit/db/mongo"
+	"github.com/ardanlabs/kit/log"
+
+	gc "github.com/patrickmn/go-cache"
+	mgo "gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+)
+
+// Contains the name of Mongo collections.
+const (
+	Collection        = "coral_items"
+	CollectionHistory = "coral_items_history"
+)
+
+// Set of error variables.
+var (
+	ErrNotFound = errors.New("Item Not found")
+)
+
+// =============================================================================
+
+// c contans a cache of set values. The cache will maintain items for one
+// second and then marked as expired. This is a very small cache so the
+// gc time will be every hour.
+
+const (
+	expiration = time.Second
+	cleanup    = time.Hour
+)
+
+var cache = gc.New(expiration, cleanup)
+
+// =============================================================================
+
+// create an item out of its type, version and data or die trying
+//  if the item already exists, adopt it's Id but do not presume to update
+func Create(context interface{}, db *db.DB, t string, v int, d ItemData) (Item, error) {
+
+	i := Item{}
+
+	// Check to see if this item already exists in the db
+	//  this is done by checking it's "foreign identity":
+	//  Checking for its type and it's Type.Id fields
+
+	// get this data's IdField value for this type
+	idValue := getDatumByKey(Types[t].IdField, d)
+
+	// if there is not a value, generate a new id and continue
+	if idValue == nil {
+		i.Id = bson.NewObjectId()
+	} else { // if there is an id, look it up
+
+		// create the field path for the id field in the d subdoc
+		dbIdField := "d." + Types[t].IdField
+
+		// build a query
+		var q = bson.M{"t": t, dbIdField: idValue}
+
+		// get one by query, assuming data consistency
+		dbItem, err := GetOneByQuery(context, db, q)
+		if err != nil {
+			return Item{}, err
+		}
+
+		// if we found an item, assign the id to the new item
+		if dbItem != nil {
+			i.Id = dbItem.Id
+		} else { // otherwise, new id it is
+			i.Id = bson.NewObjectId()
+		}
+
+	}
+
+	// validate and set type
+	if isRegistered(t) == false {
+		return i, errors.New("Type not recognized: " + t)
+	}
+	i.Type = t
+
+	// set default version if zero value
+	if v == 0 {
+		v = DefaultVersion
+	}
+	i.Version = v
+
+	// set the data into the item
+	i.Data = d
+
+	// get the relationships for this item
+	rels, err := GetRels(context, db, &i)
+	if err != nil {
+		return Item{}, err // todo, clean up empty type return
+	}
+	i.Rels = *rels
+
+	return i, nil
+}
+
+// Items are trasparently created or updated depending on thier existence
+func Upsert(context interface{}, db *db.DB, item *Item) error {
+
+	// validate our item
+	if err := item.Validate(); err != nil {
+		log.Error(context, "Upsert", err, "Completed")
+		return err
```

Comment:
Similar comment here and throughout on the errors.  There is a chance this logging could get confusing if you are logging here, and when that error is returned, and when that error is returned, etc.  Dave Cheney's errors package makes this a lot easier, as you can wrap errors and always get back to a root cause.  Something we might consider starting to utilize, although we will need to continue the discussion around errors and logging.